### ISSUE TITLE
gui.show_dialog error quits to main menu

### ISSUE THREAD

**Issue by ProditorMagnus** [wesnoth/wesnoth#9178]
### Game and System Information

- **Version:** 1.18.2
- **OS:** windows

### Description of the bug

When dialog definition is invalid, there is message like 
![image](https://github.com/user-attachments/assets/627eade8-82dd-4490-a546-b4c41060b51c)
but after confirming it, game quits to main menu.

### Steps to reproduce the behavior

```
# lua wesnoth.require("wml-utils").handle_event_commands(wml.load("~add-ons/EventLoader/action.cfg"))

[lua]
code=<<
T = wml.tag

function _G.do_nothing() end

_G.delay_dialog = {
  T.tooltip { id = "tooltip_large" },
  T.helptip { id = "tooltip_large" },
  T.grid { T.row {
    T.column { T.grid {
      T.row { T.column { T.grid { 
		T.row {
			T.column {
				T.button { id = "ok", label = "I have seen my leader now" }
			}
		}
	  } } }
    } },
    T.column { T.image { id = "the_image" } }
  } }
}


-- correct: delay_dialog[3][2][1][2][1][2][1][2][1][2][1][2][1][2]
if true then
table.insert(delay_dialog[3][2][1][2][1][2][1][2][1][2][1][2], T.row {
	T.column {
		T.label { id = "other", label = "other" }
	}
})
end

gui.show_dialog(_G.delay_dialog, do_nothing, do_nothing)

>>
[/lua]

```

### Expected behavior

Lua error in chat area, but game continues.

**Comment by AI0867** [comment]
@Vultraz You've been debugging this. Have you identified where the exception is thrown?

You've also found a spot where you can catch it with no effect (probably due to the `lua_jailbreak_exception` system), was that `intf_show_dialog` in `scripting/lua_widget_methods.cpp`?

**Comment by Vultraz** [comment]
Turns out this issue is more complicated than it seems at first. Copying my findings from Discord:

> Specifically, trying to fix https://github.com/wesnoth/wesnoth/issues/9178. the gui2 window construction functions called from intf_show_dialog (which implements the gui.show_dialog function) have various VALIDATE calls (which throw wml_deception if a condition is false, which inherits from lua_jailbreak_exception). The problem is that the validation failure causes the game to exit to main menu, rather than just continuing the game.
Part of the problem seems to be that wml_exception doesn’t inherit from std::exception, and thus its error message ain’t pushed to the Lua stack to be handled in luaW_pcall as an chat error message
(Looking at our definition of LUAI_TRY)
Now, in most cases I imagine a wml_exception should be fatal - but here, we can just throw up the error message and say we can’t build the window
There’s no reason for the function to be that fatal
We already get the error message shown, but since wml_exception is a lua_jailbreak_exception it registers itself to be rethrown outside the lua layer
lua_jailbreak_exception::store makes a copy of the exception, so even if I catch the initial validation failure in intf_show_dialog, the copy is still handled separately by the Lua exception pipeline
I suppose I could catch the exception and forcibly clear the copy … but that seems wrong

**Comment by Vultraz** [comment]
@AI0867 yes, it was `intf_show_dialog` in `scripting/lua_widget_methods.cpp`. The exception in thrown by failed `VALIDATE` conditions when constructing the `gui2::builder_window::window_resolution` object.

**Comment by AI0867** [comment]
From what I see, the current system of
* Always store a jailbreak exception on construct, if `jail_depth > 0`
* Only clear a jailbreak exception on rethrow at `jail_depth == 0`

Is a bit over a year old. This system means that jailbreak exceptions effectively *cannot be caught* as long as lua is in the call stack, as they will be rethrown as soon as control passes through lua again.

I doubt that was entirely the intention, but I also don't know whether this behaviour is now load-bearing.

So the options that I see:
1. Leave behaviour as-is: any `wml_exception` (or other `lua_jailbreak_exception`) thrown while lua is on the stack, cannot effectively be caught until the last lua stack layer has returned/unrolled.
2. Add a `catch` or `clear` method of some sort to `lua_jailbreak_exception` to explicitly mark the exception as caught, and clear it from the `rethrow` mechanism. There may be a neat trick to (nearly) automatically do this, or it might just be manual typing every time.
3. Revert the "store on construct" behaviour, and only store `lua_jailbreak_exception`s on the wesnoth/lua interface.

The relevant commit is https://github.com/wesnoth/wesnoth/commit/78d3b0c05c9834796b618b66438b88971dde30e0, which was authored by @pehjota and committed/merged by @Pentarctagon. Maybe they have some insights about the whole system.

PR https://github.com/wesnoth/wesnoth/pull/8234 has more details.

**Comment by Pentarctagon** [comment]
I have no idea how relying on LUAI_TRY prevents the use of system lua (or tbh how that commit prevents usage of it since that PR doesn't remove any usages of LUAI_TRY anyway).

**Comment by Vultraz** [comment]
Part of the problem is that `wml_exception` usually *should* be fatal, especially if it's a validation error with the core UI. The challenge here is that we're dealing with user input with all the concomitant pitfalls... I was thinking schema validation might be a good way to catch errors before the `VALIDATE` calls get hit...

**Comment by soliton-** [comment]
> I have no idea how relying on LUAI_TRY prevents the use of system lua (or tbh how that commit prevents usage of it since that PR doesn't remove any usages of LUAI_TRY anyway).

"LUAI_TRY() is an internal part of Lua that may not exist forever, and reliance on overriding it prevents the use of system copies of Lua."

You cannot modify (override) `LUAI_TRY` in system Lua since you cannot modify anything in system Lua. The mentioned commit makes it so the only thing our custom `LUAI_TRY` does over the default is convert `std::exception`s into a Lua error and assert on unknown exceptions. (Hopefully either of those things just mean more difficulty debugging but no behavior difference when things work normally.) `lua_jailbreak_exception` does no "jail breaking" in LUAI_TRY anymore but merely by being constructed.

This seems somewhat unrelated to the issue at hand though since as Vultraz says the issue is that the normal "jail break" behavior is fine but not wanted here. So that fact needs to be communicated some way so different error handling can be used in this specific case.

**Comment by hladkyjakub** [comment]
Just confirmed it in 1.19.19, may be fixed in 22/23 tho

**Comment by ProditorMagnus** [comment]
With 1.19.25, Lua console gets
```
$ wesnoth.require("wml-utils").handle_event_commands(wml.tovconfig(wml.load("~add-ons/EventLoader/action.cfg")))
Lua Error:
  When executing, Lua runtime error: ~wml:Error occured inside [lua]: std::exception
stack traceback:
	[C]: in function '.error'
	lua/core/wml.lua:244: in function 'wml.error'
	lua/wml-utils.lua:150: in field 'handle_event_commands'
	[string "debug command"]:1: in main chunk
```
Nothing is visible for player. That is misleading, player should be aware of the error to be able to report it.

**Comment by Wedge009** [comment]
Does this happen in any current core campaigns? I'm wondering about the justification for the blocker label.

**Comment by soliton-** [comment]
Should be fixed with #11044.
